libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,611 @@
+    *pCurrentState = MQTTStateNull;
+    if( packetId == MQTT_PACKET_ID_INVALID )
+    {
+        index = recordCount;
+    }
+    else
+    {
+        for( index = 0; index < recordCount; index++ )
+        {
+            if( records[ index ].packetId == packetId )
+            {
+                *pQos = records[ index ].qos;
+                *pCurrentState = records[ index ].publishState;
+                break;
+            }
+        }
+    }
+    return index;
+}
+
+static MQTTStatus_t _addRecord( MQTTPubAckInfo_t * records,
+                                size_t recordCount,
+                                uint16_t packetId,
+                                MQTTQoS_t qos,
+                                MQTTPublishState_t publishState )
+{
+    MQTTStatus_t status = MQTTNoMemory;
+    int32_t index = 0;
+    size_t availableIndex = recordCount;
+
+    assert( packetId != MQTT_PACKET_ID_INVALID );
+    assert( qos != MQTTQoS0 );
+
+    /* Start from end so first available index will be populated. */
+    for( index = ( (int32_t ) recordCount - 1 ); index >= 0; index-- )
+    {
+        if( records[ index ].packetId == MQTT_PACKET_ID_INVALID )
+        {
+            availableIndex = ( size_t ) index;
+        }
+        else if( records[ index ].packetId == packetId )
+        {
+            /* Collision. */
+            status = MQTTStateCollision;
+            availableIndex = recordCount;
+            break;
+        }
+        else
+        {
+            /* Empty else clause. */
+        }   
+    }
+
+    if( availableIndex < recordCount )
+    {
+        records[ availableIndex ].packetId = packetId;
+        records[ availableIndex ].qos = qos;
+        records[ availableIndex ].publishState = publishState;
+        status = MQTTSuccess;
+    }
+
+    return status;
+}
+
+static void _updateRecord( MQTTPubAckInfo_t * records,
+                           size_t recordIndex,
+                           MQTTPublishState_t newState,
+                           bool shouldDelete )
+{
+    if( shouldDelete )
+    {
+        records[ recordIndex ].packetId = MQTT_PACKET_ID_INVALID;
+        records[ recordIndex ].qos = MQTTQoS0;
+        records[ recordIndex ].publishState = MQTTStateNull;
+    }
+    else
+    {
+        records[ recordIndex ].publishState = newState;
+    }
+}
+
+MQTTStatus_t MQTT_ReserveState( MQTTContext_t * pMqttContext,
+                                uint16_t packetId,
+                                MQTTQoS_t qos )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    size_t recordIndex = MQTT_STATE_ARRAY_MAX_COUNT;
+    MQTTQoS_t tempQoS = MQTTQoS0;
+    MQTTPublishState_t tempState = MQTTStateNull;
+
+    recordIndex = _findInRecord( pMqttContext->outgoingPublishRecords,
+                                 MQTT_STATE_ARRAY_MAX_COUNT,
+                                 packetId,
+                                 &tempQoS,
+                                 &tempState );
+
+    if( qos == MQTTQoS0 )
```

Comment:
QOS0 records are not added. (assert in _addRecord () ), so should we even check for those. Wondering if the first two conditions can be checked before _findInRecord() to make this function more efficient.

Reply:
Although these functions are intended to be internal to the library, I didn't want to add asserts here since the functions are not static. Although these functions should not be called for QoS 0, I added this check in case the caller did not perform a check for it. The only difference is this `if` block, all other parts of the function should be the same.

Reply:
I'm just wondering if this check should happen before _findInRecord() is called.

Reply:
Now that it's brought up, I realize the call to `_findInRecord()` could just be removed. Its only purpose is collision detection which `_addRecord()` does anyway.